### DISCUSSION TITLE
Composable errors and type safety

### DISCUSSION THREAD

**Discussion by JacksonUtsch** [pointfreeco/swift-composable-architecture#1332] [Beta]
Here I am, a broken record. I'm looking for method to modularize error handling and have it be type safe. This is generally done with the `Result` type in Swift, as throws are untyped.

Why would we want type safe errors and for them to be modular? Exhaustive applications! Without such, how are we to handle unexpected error cases? Type safety allows us to understand potential errors and appropriately deal with each case as need be. We can even easily create user or developer facing associated messages.

With the new `ReducerProtocol` I considered the option of using the `Result` type as the `Action`. I thought it would work out better than it did. There are consequences of doing so, like having to handle error cases in each reducer rather than composing the reducers. It gets a little messy.

If `Error` types are opted into the `ReducerProtocol` type then composition can take place where errors can be propagated through feature domains and handled globally where some or most errors can't be handled locally in their feature domain.

[Here](https://github.com/JacksonUtsch/swift-composable-architecture/tree/protocol) is my brief testing with from protocol branch

Episode [196](https://www.pointfree.co/episodes/ep196-async-composable-architecture-tasks) talks about the error situation some

**Comment by JacksonUtsch** [comment]
If this direction is not taken, what are we supposed to do with global errors that we want to present the user? These are unable to be propagated from features.

**Reply by mbrandonw** [reply]
Dependencies are always an escape hatch to do anything you want. You could have a `ToastClient` with a `showToast` method that can be invoked from any layer in your application. And further the client expose a `toasts` async sequence that could be subscribed to at the root for displaying little banner toasts on top of everything else in the application. This would allow every feature to display global toasts without much work.

But as of right now I don't see a library-wide solution to this. It seems domain specific and up to people using the library to decide how they want to handle errors.

Also it doesn't seem right to allow reducers to `throw`. That would mean if any child reducer throws an error it will completely short-circuit any reducer that hasn't yet run, which could break a lot of expectations and invariants of the system, such as closing resources and cancelling effects.

**Reply by JacksonUtsch** [reply]
I would think utilizing a result type equivalent would be better over using `throws`. `Store` would still assume a `Never` failure case as the errors would have to be handled prior. Biggest concern as you said would be short-circuiting a reducer. I think this could be handled fine with explicit handling of error paths. Additionally you may want to handle a category of errors a certain way and others another. This composition could aid that. 

I am currently deciding on a new app whether I'd like the feature dependencies to rely on TCA or not. Previously had issues with SwiftPM conflicting versions which prevented building. I haven't seen updates to this in SwiftPM docs so I assume the behavior is the same.

If not, my type signature looks like this for reducers

```swift
public static func reducer(state: inout State, action: Action, environment: Environment) -> AsyncStream<Result<Action, Errors>> {
```

Although, if I don't depend on TCA I might not be able to get rid of the `Environment` requirement..

I'll look at using dependencies for a solution, need to think of how the app will scale up and handle these failure cases.

**Reply by fabstu** [reply]
>  Previously had issues with SwiftPM conflicting versions which prevented building

offtopic: I had the same issue when the TCA version set inside Package.swift and the TCA version imported by the Xcode workspace differed. Setting the same version in both places fixed it for me.

**Reply by JacksonUtsch** [reply]
The issue I was referring to is when I have two package dependencies that depend on different versions of TCA. But, I have come across the issue where Package.resolved differs in my workspace and xcode project as well.

The version conflict is something that package managers can implement, just download 2 of the same dependency with different versions.

**Reply by JacksonUtsch** [reply]
@mbrandonw So a dependency could be used to report errors within feature domains but this would require each feature to depend on this error dependency, creating a bad coupling when these are separate packages. Unless each used their own dependency which would create a lot of bloat. I'm not seeing a good alternative.

**Reply by mbrandonw** [reply]
> ...but this would require each feature to depend on this error dependency, creating a bad coupling when these are separate packages.

I'm not sure I see this problem. The dependency I am imagining is super lightweight. It could even be a wrapper around Notification Center, or even just use Notification Center directly in the reducer.

**Reply by JacksonUtsch** [reply]
It could be quite lightweight:

```swift
public struct Errors<T> {
	public var publisher: () -> AnyPublisher<T?, Never>
	public var report: (T) -> Void
```

Though, I find it cumbersome when every dependency must rely on this client and utilize the same error type, meaning having additional requirements would cause one to change all the errors in existing dependencies or make assumptions (optional casting) without type safety. Additionally when a `Feature` is shared among various applications as with my case, it makes any changes more tedious and errors unclear. What reducer and action did this error originate? Maybe this could be explicitly added in a complicated manner to the `Errors` client but again this could be implicit in the reducer.

There is just no way for type safety with handling errors this way and have to be based on developer knowledge.

**Reply by stephencelis** [reply]
@JacksonUtsch I mentioned it in a reply to your other recent comment about error handling, but I think concrete demonstrations are necessary to help sell and better explain your points, and/or help us find alternate solutions. Wanna sketch out a working case study or two that captures the problem at hand and potential solutions?

**Reply by JacksonUtsch** [reply]
@stephencelis Agreed. I'm finding it quite difficult to make all the changes necessary for the library itself with the intricacies of task execution and while supporting `Publisher`. Would an independent project be suitable for demonstration or perhaps if you have a suggestion?

**Reply by JacksonUtsch** [reply]
@stephencelis With some experimenting it seems `Action` can be used to adopt a failure case. This does change the behavior of streams continuing to execute but ultimately this may be preferred. So no changes are needed with this method! Thank you for entertaining the idea, I may create a case study for insight on the use case.

**Reply by JacksonUtsch** [reply]
[Here](https://github.com/JacksonUtsch/swift-composable-architecture/blob/error-handling/Examples/CaseStudies/SwiftUICaseStudies/Errors.swift) is a case study I made that demonstrates the idea and some use cases. +1 from me to add `_Observe` to TCA

**Reply by JacksonUtsch** [reply]
@stephencelis Not sure if you've been able to take a look. Does this method of propagating errors merit any validity?
